### PR TITLE
drivers: serial: nrfx: Skip poll out when device is not active

### LINKED ISSUE #30230: printk and power management incompatibility
**Describe the bug**
Software enters infinite loop when uart assigned to the console enters OFF state.
The driver ends in an infinite loop waiting for sending to finish.
Confirmed on nRF52 devices, not sure if other drivers behave the same.

**To Reproduce**
 1. Enable CONFIG_LOG_MINIMAL in the project config.
 2. Enter power off state on the uart:
```
	const struct device *cons = device_get_binding(CONSOLE_LABEL);
	device_set_power_state(cons, DEVICE_PM_OFF_STATE, NULL, NULL);
```
 3. Log anything, for example using NET_LOG function.
 or 
 try coap_client sample in NCS with overlay-mtd.conf applied and latest zephyr.

**Expected behavior**
Logs are not shown, but the application keeps working correctly

**Observed behavior**
Application stays in an infinite loop on sending the logs. 

**Impact**
Blocker - Thread Sleepy End Device sample is not working at all. It is not possible to disable minimal log at runtime.

**Logs and console output**
N/A - no logs as uart is disabled

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: gcc-arm-none-eabi-9-2019-q4
 - board: nrf52840dk_nrf52840
 - commit: ??

**Additional context**
Function that causes an infinite loop: zephyr/drivers/serial/uart_nrfx_uarte.c:uarte_nrfx_poll_out

### PR DESCRIPTION
**Pull request by nordic-krch**
Added early return from uart_poll_out when uart device is not in active
state. Poll_out is used from many contexts by multiple users and currently
there is no mechanism in power management to disable them (console, logger)
when going to inactive power state.

Additionally, moved setting new state when disactivating the device to ensure
that no poll_out will happen after disabling the UARTE.

Fixes #30230.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Review by anangl — APPROVED** [review]

**Review by Mierunski — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

File: drivers/serial/uart_nrfx_uarte.c [hunk 1]
```diff
@@ -1534,6 +1539,8 @@ static void uarte_nrfx_set_power_state(const struct device *dev,
 			NRF_UARTE_PSEL_DISCONNECTED) {
 			nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
 		}
+
+		data->pm_state = new_state;
```
Comment (MarekPorwisz) [inline drivers/serial/uart_nrfx_uarte.c:1543]: Shouldn't this be inside an ifdef? I think that data->pm_state is only available when CONFIG_DEVICE_POWER_MANAGEMENT is set

File: drivers/serial/uart_nrfx_uarte.c [hunk 2]
```diff
@@ -1546,6 +1553,8 @@ static void uarte_nrfx_set_power_state(const struct device *dev,
 			return;
 		}
 
+		data->pm_state = new_state;
```
Comment (MarekPorwisz) [inline drivers/serial/uart_nrfx_uarte.c:1556]: same

File: drivers/serial/uart_nrfx_uarte.c (hunk 1 above)
Comment (nordic-krch) [reply drivers/serial/uart_nrfx_uarte.c:1543]: whole function is conditionally compiled

Comment (MarekPorwisz) [reply drivers/serial/uart_nrfx_uarte.c:1543]: ahh. sorry, folding again. it showed like it is a part of uarte_nrfx_poll_out function.

**Pull request merged**